### PR TITLE
Add iterator protocol for modules

### PR DESCRIPTION
**Pull request by kdheepak**
**Usage**:

```

julia> using OpenDSSDirect

julia> dss("""
       clear
       redirect ./examples/8500-Node/Run_8500Node.dss
       """)
""

julia> for i in Lines
           @show typeof(i), Lines.Name()
           if i >= 5
               break
           end
       end
(typeof(i), Lines.Name()) = (Int64, "hvmv_sub_connector")
(typeof(i), Lines.Name()) = (Int64, "ln5502549-1")
(typeof(i), Lines.Name()) = (Int64, "ln6259988-1")
(typeof(i), Lines.Name()) = (Int64, "ln6077796-1")
(typeof(i), Lines.Name()) = (Int64, "ln5835135-2")
```

Difference in traceback for the user when iterating on modules in other packages:

```
$ julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.1.0 (2019-01-21)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> #########################################################

julia> # default stack trace

julia> using LinearAlgebra

julia> for i in LinearAlgebra
       @show i
       end
ERROR: MethodError: no method matching iterate(::Module)
Closest candidates are:
  iterate(::Core.SimpleVector) at essentials.jl:568
  iterate(::Core.SimpleVector, ::Any) at essentials.jl:568
  iterate(::ExponentialBackOff) at error.jl:199
  ...
Stacktrace:
 [1] top-level scope at ./none:0

julia> #########################################################

julia> # stacktrace after `using OpenDSSDirect`

julia> using OpenDSSDirect

julia> for i in Lines
           @show i
       end

julia> using LinearAlgebra

julia> for i in LinearAlgebra
       @show i
       end
ERROR: MethodError: no method matching iterate(::Module)
Closest candidates are:
  iterate(::Module) at /Users/$USER/GitRepos/OpenDSSDirect.jl/src/iterators.jl:3
  iterate(::Module, ::Int64) at /Users/$USER/GitRepos/OpenDSSDirect.jl/src/iterators.jl:17
  iterate(::Core.SimpleVector) at essentials.jl:568
  ...
Stacktrace:
 [1] iterate(::Module) at /Users/$USER/GitRepos/OpenDSSDirect.jl/src/iterators.jl:4
 [2] top-level scope at ./none:0

julia>
```

### CODE REVIEW

**Comment by tshort** [comment]
I'm against this. This is "type piracy", meaning that you're defining methods for types you don't own. This is strongly frowned upon in the Julia community because it can cause packages to interact badly. One way around this is to use a wrapper type and define the iterator for the wrapper type. Here's an example of how a wrapper struct `EachMember` could be used:

```julia
for i in EachMember(CapControls)
    ...
end
```

This iterator is also a bit odd in that the value returned is just an index. Normally, an iterator returns an object to act upon. I don't have a suggestion for this. It's just an artifact on how the OpenDSS interface works. The iterator modifies state that is stored in OpenDSS.

**Comment by kdheepak** [comment]
It did feel odd and a little dirty, even after I added the method error for this. I'll close this and add the iterator on a wrapper instead, similar to how I do it in Python.

Do you have preferences between the following?

**Option a**

```
for n in eachmember(Lines, :Name)
    @show n
end
```

**Option b**

```
for n in eachmember(Lines, Lines.Name)
    @show n
end
```

**Comment by tshort** [comment]
Why not just `eachmember(Lines)`?

**Comment by kdheepak** [comment]
Because as you mention, it will just returns an index. I can add that as well. Both **option a** and **option b** will return the object that is returned by the function call.

**Comment by kdheepak** [comment]
Closing, will reopen when I have an implementation ready for review.

**Comment by kdheepak** [comment]
@tshort If this looks good to you, I'll add more tests and docs and merge.

```julia
arr = String[]
for i in OpenDSSDirect.EachMember(Lines); 
    push!(arr, Lines.Name()); 
end
for (i, n) in enumerate(OpenDSSDirect.EachMember(Lines, Lines.Name))
    @test n == arr[i]
end
```

File: src/iterators.jl [hunk 1]
```diff
@@ -0,0 +1,44 @@
+
+struct EachMember
+    m::Module
+    f::Union{Nothing, Function}
```
Comment (tshort) [inline src/iterators.jl]: I don't understand what `f` is for. In the `iterate` definitions below, I suspect that `f` is always `nothing`. You may be able to take out some code. Note that I didn't run it to confirm.

Comment (tshort) [reply src/iterators.jl]: I get it now. It's for `enumerate(OpenDSSDirect.EachMember(Lines, Lines.Name))`.

Comment (tshort) [reply src/iterators.jl]: It's a bit of an unusual iterator, so docs will help here.

**Comment by PMeira** [comment]
> The Python interface uses the latter, i.e. **Option b**, see [here](https://github.com/dss-extensions/OpenDSSDirect.py/blob/feb5ab14b46352ffbeac8b45751dc28c9148a075/opendssdirect/utils.py#L16-L32) for implementation.

For dss_python, since it uses classes instead of modules, I used a [simple generator](https://github.com/dss-extensions/dss_python/blob/bacb78ed41fc7c0229b51534ce4f7adbadf4b983/dss/dss_capi_ir/ILines.py#L315):

```python
    def __iter__(self):
        idx = self.First
        while idx != 0:
            yield self
            idx = self.Next
```

I was going to propose that we change ODD.py to achieve the same.

For buses, we don't have the First/Next iteration pattern but I added `Bus_Get_Next` to simplify it:

```python
    def __iter__(self):
        n = self._lib.Circuit_SetActiveBusi(0)
        while n == 0:
            yield self
            n = self._lib.Bus_Get_Next()
```

**Comment by kdheepak** [comment]
@PMeira are suggesting changing all the modules to classes in ODD.py? We should open an issue over there to discuss this further.

**Comment by kdheepak** [comment]
I've added the following docstring:

___

    EachMember(m::Module)
    EachMember(m::Module, f::Function)

Call function `f` from module `m` for each member or each element in module in a iterator fashion.

```julia
julia> for n in OpenDSSDirect.EachMember(Lines, Lines.Name)
            @show n
       end
n = "hvmv_sub_connector"
n = "ln5502549-1"
n = "ln6259988-1"
n = "ln6077796-1"
n = "ln5835135-2"
...
```

This also works without a function. By default, it returns the index of the elment.

```julia
julia> for i in OpenDSSDirect.EachMember(Lines)
           @show i, Lines.Name()
       end
(i, Lines.Name()) = (1, "hvmv_sub_connector")
(i, Lines.Name()) = (2, "ln5502549-1")
(i, Lines.Name()) = (3, "ln6259988-1")
(i, Lines.Name()) = (4, "ln6077796-1")
(i, Lines.Name()) = (5, "ln5835135-2")
...
```

In each iteration of the body of the `for` loop a new element of the module can be accessed.

File: src/iterators.jl (hunk 1 above)
Comment (kdheepak) [reply src/iterators.jl]: Added docs in 65cfc5f

**Comment by kdheepak** [comment]
Tests all pass, so I'll go ahead and merge this.

**Pull request merged**